test/integration/sds_dynamic_integration_test.cc

```diff
@@ -0,0 +1,345 @@
+
+      // Add a static sds cluster
+      auto* sds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      sds_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+      sds_cluster->set_name("sds_cluster");
+      sds_cluster->mutable_http2_protocol_options();
+    });
+
+    HttpIntegrationTest::initialize();
+    client_ssl_ctx_ = createClientSslTransportSocketFactory(false, false, context_manager_);
+  }
+
+  void createUpstreams() override {
+    HttpIntegrationTest::createUpstreams();
+    // SDS upstream
+    fake_upstreams_.emplace_back(
+        new FakeUpstream(0, FakeHttpConnection::Type::HTTP2, version_, enable_half_close_));
+  }
+
+  void TearDown() override {
+    cleanUpSdsConnection();
+
+    client_ssl_ctx_.reset();
+    cleanupUpstreamAndDownstream();
+    fake_upstream_connection_.reset();
+    codec_client_.reset();
+  }
+
+  Network::ClientConnectionPtr makeSslClientConnection() {
+    Network::Address::InstanceConstSharedPtr address = getSslAddress(version_, lookupPort("http"));
+    return dispatcher_->createClientConnection(address, Network::Address::InstanceConstSharedPtr(),
+                                               client_ssl_ctx_->createTransportSocket(), nullptr);
+  }
+
+private:
+  Network::TransportSocketFactoryPtr client_ssl_ctx_;
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, SdsDynamicDownstreamIntegrationTest,
+                        GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// A test that SDS server send a good server secret for a static listener.
+// The first ssl request should be OK.
+TEST_P(SdsDynamicDownstreamIntegrationTest, BasicSuccess) {
+  pre_worker_start_test_steps_ = [this]() {
+    createSdsStream(*(fake_upstreams_[1]));
+    sendSdsResponse(getServerSecret());
+  };
+  initialize();
+
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection();
+  };
+  testRouterHeaderOnlyRequestAndResponse(true, &creator);
+}
+
+// A test that SDS server send a bad secret for a static listener,
+// The first ssl request should fail at connecting.
+// then SDS send a good server secret,  the second request should be OK.
+TEST_P(SdsDynamicDownstreamIntegrationTest, WrongSecretFirst) {
+  pre_worker_start_test_steps_ = [this]() {
+    createSdsStream(*(fake_upstreams_[1]));
+    sendSdsResponse(getWrongSecret());
+  };
+  initialize();
+
+  codec_client_ = makeRawHttpConnection(makeSslClientConnection());
+  // the connection state is not connected.
+  EXPECT_FALSE(codec_client_->connected());
+  codec_client_->connection()->close(Network::ConnectionCloseType::NoFlush);
+
+  sendSdsResponse(getServerSecret());
+
+  // Wait for ssl_context_updated_by_sds counter.
+  if (version_ == Network::Address::IpVersion::v4) {
+    test_server_->waitForCounterGe(
+        "listener.127.0.0.1_0.server_ssl_socket_factory.ssl_context_update_by_sds", 1);
```

Comment:
It would be nice to be able to set the stat_prefix on listeners, so that you don't have to care which ip version the test is for. But that's a fix for another day.

Reply:
Good point. Thanks!